File: scripts/reference-solutions/stable-trading-windows.py
import sys
from collections import deque


def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    k = int(data[1])
    prices = list(map(int, data[2 : 2 + n]))

    max_dq = deque()  # decreasing: front is window max
    min_dq = deque()  # increasing: front is window min
    left = 0
    ans = 0
    for right in range(n):
        value = prices[right]
        while max_dq and prices[max_dq[-1]] <= value:
            max_dq.pop()
        max_dq.append(right)
        while min_dq and prices[min_dq[-1]] >= value:
            min_dq.pop()
        min_dq.append(right)

        while prices[max_dq[0]] - prices[min_dq[0]] > k:
            left += 1
            if max_dq[0] < left:
                max_dq.popleft()
            if min_dq[0] < left:
                min_dq.popleft()

        ans += right - left + 1
    print(ans)


main()

File: scripts/reference-solutions/stable-trading-windows.cpp
#include <deque>
#include <iostream>
#include <vector>
using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  long long n, k;
  cin >> n >> k;
  vector<long long> prices(n);
  for (long long& value : prices) {
    cin >> value;
  }

  deque<long long> maxDq;  // decreasing: front is window max
  deque<long long> minDq;  // increasing: front is window min
  long long left = 0;
  long long ans = 0;
  for (long long right = 0; right < n; ++right) {
    long long value = prices[right];
    while (!maxDq.empty() && prices[maxDq.back()] <= value) {
      maxDq.pop_back();
    }
    maxDq.push_back(right);
    while (!minDq.empty() && prices[minDq.back()] >= value) {
      minDq.pop_back();
    }
    minDq.push_back(right);

    while (prices[maxDq.front()] - prices[minDq.front()] > k) {
      ++left;
      if (maxDq.front() < left) {
        maxDq.pop_front();
      }
      if (minDq.front() < left) {
        minDq.pop_front();
      }
    }

    ans += right - left + 1;
  }

  cout << ans << '\n';
  return 0;
}
